include/ginkgo/core/base/array.hpp

```diff
@@ -294,14 +298,36 @@ class Array {
             this->resize_and_reset(0);
             return *this;
         }
-        this->resize_and_reset(other.get_num_elems());
-        exec_->copy_from(other.get_executor().get(), num_elems_,
+        // We allow resizing only if we are an array, we cannot resize a view!
+        // If we are a view, we do bound checking to ensure the data can be
+        // copied over.
+        if (data_.get_deleter().target_type() != typeid(view_deleter)) {
+            this->resize_and_reset(other.get_num_elems());
+        } else {
+            GKO_ENSURE_COMPATIBLE_BOUNDS(other.get_num_elems(),
+                                         this->num_elems_);
+        }
+        exec_->copy_from(other.get_executor().get(), other.get_num_elems(),
                          other.get_const_data(), this->get_data());
         return *this;
     }
 
     /**
-     * Moves data from another array.
+     * Moves data from another array or view. Only the pointer and deleter type
+     * change, a copy only happens when targeting another executor's data. This
+     * means that in the following situation:
+     * ```cpp
+     *   gko::Array<int> a; // an existing array or view
+     *   gko::Array<int> b; // an existing array or view
+     *   b = std::move(a);
+     * ```
+     * Depending on whether `a` and `b` are array or view, this happens:
+     * + `a` and `b` are view, `b` becomes the only valid view of `a`
+     * + `a` and `b` are array, `b` becomes the only valid array of `a`
+     * + `a` is a view and `b` is an array, `b` frees its data and becomes the
+     *    only valid view of `a`
+     * + `a` is an array and `b` is a view, `b` becomes the only valid array
+     *    of `a`
```

Comment:
I think it would be nice to mention that the pointer of `a` will be a `nullptr` (or, if we want to be vague and closer to the standard description: `a is in a valid but unspecified state afterwards`) after the move is done (in every case, right?).
I know that you try to say that with the "only valid [view, array] of `a`", but I expect that others might not understand the implications of that.